File: offer56-I-shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-lcof/python/Solution.py
class Solution:
    def singleNumbers(self, nums: List[int]) -> List[int]:
        xorResult = 0
        n = len(nums)
        for i in range(n):
            xorResult ^= nums[i]
        tag = 1
        while xorResult & tag == 0:
            tag = tag << 1
        a = 0
        b = 0
        for i in range(n):
            if (nums[i] & tag) == 0:
                a ^= nums[i]
            else:
                b ^= nums[i]
        return [a,b]

File: offer56-I-shu-zu-zhong-shu-zi-chu-xian-de-ci-shu-lcof/cpp/Solution.cpp
class Solution {
public:
    vector<int> singleNumbers(vector<int>& nums) {
        int xorResult = 0;
        int n = nums.size();
        for (int i = 0; i < n; ++i) {
            xorResult ^= nums[i];
        }
        int tag = 1;
        while ((xorResult & tag) == 0) {
            tag = tag << 1;
        }
        int a = 0;
        int b = 0;
        for (int i = 0; i < n; ++i) {
            if ((nums[i] & tag) == 0) {
                a ^= nums[i];
            } else {
                b ^= nums[i];
            }
        }
        return {a, b};
    }
};
